element_calcium_imaging/scan.py

```diff
@@ -613,3 +607,125 @@ def estimate_nd2_scan_duration(nd2_scan_obj):
+
+
+@schema
+class ScanQualityMetrics(dj.Computed):
+    """
+    Metrics to assess quality of Scan.
+    """
+
+    definition = """
+    -> ScanInfo.Field
+    """
+
+    class FrameMetrics(dj.Part):
+        definition = """
+        -> master
+        channel: int
+        ---
+        min_intensity: longblob   # Minimum value of each frame
+        mean_intensity: longblob  # Mean value at each frame
+        max_intensity: longblob   # Maximum value at each frame
+        contrast: longblob        # Contrast of each frame, difference of 99 and 1 percentiles.
+        """
+
+    class QuantalSize(dj.Part):
+        """
+        Quantities inferred from Photon Transfer Curve.
+        """
+
+        definition = """ 
+        -> master
+        channel: int
+        ---
+        min_intensity   : int      # Minimum value in movie.
+        max_intensity   : int      # Maximum value in movie.
+        quantal_size    : float    # Quantal size (variance/mean intensity slope), gain.
+        zero_level      : int      # Level corresponding to zero (computed from variance dependence)
+        quantal_frame   : longblob # Average frame expressed in quanta
+        """
+
+    def make(self, key):
+        acq_software, nchannels = (Scan * ScanInfo & key).fetch1(
+            "acq_software", "nchannels"
+        )
+
+        self.insert1(key)
+
+        for channel in range(nchannels):
+            if acq_software == "ScanImage":
+                import scanreader
+
+                movie = scanreader.read_scan(get_scan_image_files(key))[
+                    key["field_idx"], :, :, channel, :
+                ].transpose(2, 0, 1)
+            elif acq_software == "Scanbox":
+                import sbxreader
+
+                movie = sbxreader(get_scan_box_files(key))[
+                    :, key["field_idx"], channel, :, :
+                ]
+            elif acq_software == "NIS":
+                import nd2
+
+                nd2_file = nd2.ND2File(get_nd2_files(key)[0])
+
+                get_dim_pos = {k: i for i, k in enumerate(nd2_file.sizes)}
+
+                known_fields = ["T", "Z", "C", "Y", "X"]
+                fields_in_data = [x for x in known_fields if x in get_dim_pos]
+                assert len(fields_in_data) == len(
+                    get_dim_pos
+                ), f"Unknown dimension is detected in {get_dim_pos}"
+                movie = nd2_file.asarray().transpose(
+                    [get_dim_pos[x] for x in fields_in_data]
+                )
+
+                if "T" not in fields_in_data:
+                    movie = np.expand_dims(movie, 0)
+                if "Z" not in fields_in_data:
+                    movie = np.expand_dims(movie, 1)
+                if "C" not in fields_in_data:
+                    movie = np.expand_dims(movie, 2)
+
+                movie = movie[:, key["field_idx"], channel, :, :]
+
+            flat_movie = movie.reshape(movie.shape[0], -1)
```

Comment:
Reshaping does not make things simpler. You can keep the original dimensions.

Reply:
Right. I need to use axis keyword more often. Fixing it now.

Reply:
Fixed it.